### PR TITLE
Create Whats new in C# 7.1

### LINKED ISSUE #2794: Create "What's new in C# 7.1" topic
C# 7.1 is a point release whose new language features are speced out [here](https://github.com/dotnet/csharplang/blob/master/Language-Version-History.md#c-71-visual-studio-2017-version-153)

Briefly, they are:

- Async Main
- Default Expressions
- Reference Assemblies
- Inferred Tuple Element names
- Pattern Matching with Generics.
- The new <langVersion> node in csproj needs to be explained for different language features.

In addition, the "What's New in C#" landing page should be updated with information on C# 7.1.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #2794

Note that this PR is dependent on PR #2835

/cc @Pilchie @jcouv @DustinCampbell

### CODE REVIEW

File: docs/csharp/whats-new/index.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/index.md]: "This covers C# 7" -> "This covers C# 7.1" ?

File: docs/csharp/whats-new/csharp-7-1.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/csharp-7-1.md]: Not sure if you're interested in feedback on the actually copy, but the last four sentences in this paragraph all start with "This", which reads a bit awkwardly.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/csharp-7-1.md]: And again. :smile:

File: docs/csharp/whats-new/csharp-7-1.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/csharp-7-1.md]: `async Main` is not restricted to console applications.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/csharp-7-1.md]: Is this the correct name of the feature? I was under the impression they were called "default literal expressions"

File: docs/csharp/whats-new/csharp-7-1.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/csharp-7-1.md]: "version"

**Comment by DustinCampbell** [comment]
Thanks for including me @BillWagner. I did a quick read-through and left some feedback.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (abelbraaksma) [inline docs/csharp/whats-new/csharp-7-1.md]: shwon > shown

File: docs/csharp/whats-new/csharp-7-1.md
Comment (abelbraaksma) [inline docs/csharp/whats-new/csharp-7-1.md]: each features > each feature

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: I think it might be useful to include a full <PropertyGroup> element including the Configuration attribute. That helps to make it clear that this is going to be a per-configuration setting.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: "langVersion" --> `LangVersion`

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: I think that the important thing to stress here is not that it's typically the same, but that you need to also manually do it for each.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: Where previously --> Previously

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: `Main` method --> a `Main` method

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: perhaps use an `<xref>` instead of fenced Task?

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: There's a lot of repetition in this paragraph, starting with an unspecified topic on default value expressions, with (plural) topics and a link to a single topic on the default keyword, and then the reference to default value expressions.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: Many times that --> Many times when

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: about the details for this feature --> about this feature

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: There are two new compiler options that --> Two new compiler options
reference only --> reference-only

**Review by rpetrusha — CHANGES_REQUESTED** [review]

File: docs/csharp/whats-new/index.md
Comment (BillWagner) [inline docs/csharp/whats-new/index.md]: Also added .NET Core 2.0

File: docs/csharp/whats-new/csharp-7-1.md
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-7-1.md]: Right. I re-worked the sentence a bit. Yes, this new feature is "default literal expressions", which apply in the context of "default value expressions". I'm linking to updates to the topic on "default value expressions". Let me know what you think of the revision

File: docs/csharp/whats-new/csharp-7-1.md
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-7-1.md]: I looked in the about box in Visual Studio. "Version" is spelled out. Updated.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (DustinCampbell) [inline docs/csharp/whats-new/csharp-7-1.md]: I think it's fine. It _is_ a bit weird to see "default value expressions" in a list of new features.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-7-1.md]: Let me know what you think of the updated version.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-7-1.md]: I rewrote the paragraph. Let me know what you think.

**Review by DustinCampbell — APPROVED** [review]
LGTM from my perspective

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: right-hand

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: These -> These expressions

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: not sure if you wanna be future-proof here and say "and later versions" for the VS and SDK statements or "Starting with..."

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: To enable *the* ?

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: bold for UI options

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: below -> following

inline code latest so it doesn't get translated?

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: langVersion -> LangVersion? inline code?

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: know -> now

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: nit: does not -> doesn't

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: should we standardize on always using `<xref:uid>`?

**Review by mairaw — CHANGES_REQUESTED** [review]
Looks great! I have a couple of minor edits suggestions and comments.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: "langVersion" --> the `LangVersion`

**Review by rpetrusha — COMMENTED** [review]
Along with Maira's comments, I've added one. Otherwise, you've done a good a good job of addressing my previous comments. This is a really good overview, @BillWagner.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-7-1.md]: That works. I wasn't sure which was preferred.

**Comment by BillWagner** [comment]
@rpetrusha @mairaw Thanks for the 🦅 👀 review.  I addressed all your comments.  Let me know what you think.

File: docs/csharp/whats-new/csharp-7-1.md
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-1.md]: nit: the `LangVersion` element

**Review by rpetrusha — APPROVED** [review]
@BillWagner, except for one small nit, LGTM.

**Review by jcouv — APPROVED** [review]
LGTM
FYI @gafter @jaredpar

File: docs/csharp/whats-new/csharp-7-1.md
Comment (jcouv) [inline docs/csharp/whats-new/csharp-7-1.md]: Technically that's not a language feature (it's just a compiler feature), but I don't think it's worthwhile to change.

**Comment by mairaw** [comment]
I'll review again but builds has warnings that need to be addressed (e.g. broken image links)

File: docs/csharp/whats-new/csharp-7-1.md
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-1.md]: nit: this link behaves kind of weird because of the inline code in the middle. 
image

also this is missing punctuation.

**Review by mairaw — APPROVED** [review]
Looks good overall. I've left a nit comment plus the new topic is missing from the TOC. After that, it's ready to merge.

**Comment by BillWagner** [comment]
Made final updates. I'll merge once #2835 is merged, and the build passes when those topics are live.

**Comment by rpetrusha** [comment]
@BillWagner, the build completed with 3 warnings for the Getting Started topic, but for some reason the PR didn't register the build completion. I'm closing and reopening the PR.

**Comment by BillWagner** [comment]
@rpetrusha Some of the warnings will be fixed when #2835 is merged. I've addressed all the feedback on that PR. Can you approve that one and I'll merge it?

**Comment by rpetrusha** [comment]
@BillWagner, I approved it a while ago. We're just waiting on the build to complete -- the 3 warnings that I noted were from the OPS build report, but the build and this PR weren't in sync for some reason, and this PR showed the build as still in progress. I just restarted the build a few minutes ago, so once it's done, you should be able to merge.

**Comment by BillWagner** [comment]
build is stuck. closing and reopening. 😭

**Pull request merged**
